src/context/TransactionsProvider/TransactionsProvider.tsx

```diff
@@ -47,90 +51,107 @@ export const TransactionsProvider = ({ children }: TransactionsProviderProps): J
     [accountSpecifiers],
   )
 
+  /**
+   * tx history unsubscribe and cleanup logic
+   */
+  useEffect(() => {
+    // account specifiers changing will trigger this effect
+    // we've disconnected/switched a wallet, unsubscribe from tx history and clear tx history
+    if (txIds.length) {
+      console.info('TransactionsProvider: unsubscribing from tx history')
+      supportedChains.forEach(chain => chainAdapterManager.byChain(chain).unsubscribeTxs())
+      setIsSubscribed(false)
+      dispatch(txHistory.actions.clear())
+    }
+    // txIds are changed by this effect, don't cause infinite loop
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [accountSpecifiers, dispatch, chainAdapterManager, supportedChains])
+
+  /**
+   * tx history subscription logic
+   */
   useEffect(() => {
     if (!wallet) return
     if (isEmpty(assets)) return
-    ;(async () => {
-      for (const chain of supportedChains) {
-        const adapter = chainAdapterManager.byChain(chain)
-        const chainId = adapter.getCaip2()
-        if (!walletSupportChain({ chainId, wallet })) continue
-
-        const asset = Object.values(assets).find(asset => asset.caip2 === chainId)
-        if (!asset) {
-          throw new Error(`asset not found for chain ${chain}`)
-        }
+    if (!isPortfolioLoaded) return // wait for all chain portfolios to be loaded before subscribing
+    if (isSubscribed) return // don't resubscribe
+    ;(async () =>
+      Promise.all(
+        supportedChains
```

Comment:
should we expose `supportedChainIds` instead?

Reply:
yes - see PR overview - we need this from chainadapters.